### PR TITLE
General improvements to the network module

### PR DESCRIPTION
**Pull request by haarts**
This change set:
- improves the visibility signaling of the Network class
- renames and expands the `cb2` method
- reduces the coupling between the network module and the verifier and synchronizer module
- various small styling issues are fixed

### CODE REVIEW

File: lib/network.py [hunk 1]
```diff
@@ -1233,59 +1258,116 @@ def broadcast_transaction(self, transaction, callback=None):
...
         command = 'blockchain.headers.subscribe'
-        invocation = lambda c: self.send([(command, [True])], c)
+        invocation = lambda c: self._send([(command, [True])], c)
 
-        return Network.__with_default_synchronous_callback(invocation, callback)
+        return Network.__with_default_synchronous_callback(
+            invocation,
+            callback)
 
     def subscribe_to_address(self, address, callback=None):
```
Comment (SomberNight) [inline lib/network.py]: you might as well delete this method
(as it uses `blockchain.address.subscribe`)

File: lib/network.py [hunk 2]
```diff
@@ -613,58 +622,58 @@ def process_response(self, interface, response, callbacks):
             interface.server_version = result
         elif method == 'blockchain.headers.subscribe':
             if error is None:
-                self.on_notify_header(interface, result)
+                self._on_notify_header(interface, result)
             else:
                 # no point in keeping this connection without headers sub
                 self.connection_down(interface.server)
```
Comment (SomberNight) [inline lib/network.py]: _connection_down

File: lib/network.py [hunk 3]
```diff
@@ -1058,26 +1076,27 @@ def init_headers_file(self):
...
+        self._stop_network()
         self.on_stop()
 
-    def on_notify_header(self, interface, header_dict):
+    def _on_notify_header(self, interface, header_dict):
         try:
             header_hex, height = header_dict['hex'], header_dict['height']
         except KeyError:
             # no point in keeping this connection without headers sub
             self.connection_down(interface.server)
```
Comment (SomberNight) [inline lib/network.py]: _connection_down

Comment (SomberNight) [reply lib/network.py]: there's also one at (new) line 941, but it's too far from changed lines so I can't comment there

File: lib/network.py [hunk 4]
```diff
@@ -236,8 +236,15 @@ def __init__(self, config=None):
...
         self.socket_queue = queue.Queue()
-        self.start_network(deserialize_server(self.default_server)[2],
-                           deserialize_proxy(self.config.get('proxy')))
+        self._start_network(deserialize_server(self.default_server)[2],
+                            deserialize_proxy(self.config.get('proxy')))
+
+        # __hash2address contains a mapping between scripthashes and addresses.
+        # An address can be converted to a scripthash but not the other way
+        # around. Some internal method calls pass addresses but the remote
+        # understands only scriphashes, a conversion is done pre- and
```
Comment (SomberNight) [inline lib/network.py]: typo in "scripthashes"

File: lib/websockets.py [hunk 5]
```diff
@@ -96,28 +104,26 @@ def run(self):
...
-                scripthash = bitcoin.address_to_scripthash(addr)
                 self.network.get_balance_for_scripthash(
                         scripthash, self.response_queue.put)
             elif method == 'blockchain.scripthash.get_balance':
-                scripthash = r.get('params')[0]
-                addr = self.network.h2addr.get(scripthash, None)
+                addr = self.hash2address.get(scripthash, None)
                 if addr is None:
                     util.print_error(
                         "can't find address for scripthash: %s" % scripthash)
```
[the remark this answers is not in the document]
Comment (haarts) [reply lib/websockets.py]: I was on the fence about this one. I thought maybe for error reporting to the user, I didn't want to mess with that. But I too prefer it removed. Will do.

File: lib/network.py (hunk 2 above)
Comment (haarts) [reply lib/network.py]: got them all. thanks!

File: lib/websockets.py [hunk 6]
```diff
@@ -70,22 +73,27 @@ def make_request(self, request_id):
...
-                addr, amount = self.make_request(request_id)
-            except:
+                scripthash, amount = self.make_request(request_id)
+            except OSError:
                 continue
-            l = self.subscriptions.get(addr, [])
-            l.append((ws, amount))
-            self.subscriptions[addr] = l
-            self.network.subscribe_to_addresses([addr], self.response_queue.put)
+            except json.JSONDecodeError:
```
[the remark this answers is not in the document]
Comment (haarts) [reply lib/websockets.py]: will do

File: lib/verifier.py [hunk 7]
```diff
@@ -56,9 +55,12 @@ def run(self):
 
             header = blockchain.read_header(tx_height)
             if header is None:
-                index = tx_height // 2016
-                if index < len(blockchain.checkpoints):
-                    self.network.request_chunk(interface, index)
+                # Retreive headers when the transaction height is before the
```
[the remark this answers is not in the document]
Comment (haarts) [reply lib/verifier.py]: will fix

File: lib/network.py (hunk 4 above)
Comment (haarts) [reply lib/network.py]: got it

**Comment by ysangkok** [comment]
Should `scripts/server_versions` still be there? I don't see where the ElectrumX is defined.

**Comment by haarts** [comment]
No, it shouldn't. I'll remove it.

**Comment by haarts** [comment]
Is there anything else I can do?

**Pull request closed without merging**